core/primitives/piece/impl/piece_data.cpp

```diff
@@ -8,11 +8,26 @@
 #include <boost/filesystem.hpp>
 
 namespace fc::primitives::piece {
-  PieceData::PieceData(const std::string &path_to_file, int flags) {
+  PieceData::PieceData(const std::string &path_to_file, int flags)
+      : is_null_data_(false) {
     fd_ = open(path_to_file.c_str(), flags, 0644);
   }
 
+  PieceData::PieceData(PieceData &&other) noexcept
+      : fd_(kUnopenedFileDescriptor), is_null_data_(other.is_null_data_) {
+    fd_ = other.fd_;
+    other.fd_ = kUnopenedFileDescriptor;
+    other.is_null_data_ = true;
```

Comment:
```suggestion
  PieceData::PieceData(PieceData &&other) noexcept {
    std::swap(fd_, other.fd_);
    std::swap(is_null_data_, other.is_null_data_);
```

Reply:
`std::swap(is_null_data_, other.is_null_data_);` is not correct, because after this call other.is_null_data_ should be true.

Reply:
If our default constructor uses `is_null_data_=true`, then it will swap to other, and `other` will have `true` like you want.

Reply:
But when we call move constructor, default constructor doesn't call.